R/mwrite.R

```diff
@@ -0,0 +1,415 @@
+  } 
+}
+
+mwrite_model_to_list <- function(x) {
+  l <- list()
+  # Header
+  l$source <- "mrgsolve::mwrite"
+  l$mrgsolve <- as.character(packageVersion("mrgsolve"))
+  l$format <- "list"
+  l$version <- 1
+  l$model <- x@model
+  # Problem
+  l$prob <- ""
+  # Critical items
+  l$param <- as.list(param(x))
+  l$init <- as.list(init(x))
+  l$capture <- character(0)
+  if(length(x@capture)) {
+    capture <- names(x@capture)
+    new_names <- unname(x@capture)
+    same_name <- capture == new_names
+    l$capture <- paste0(new_names, " = ", capture)
+    l$capture[same_name] <- capture[same_name]
+  }
+  # Omega and Sigma
+  l$omega <- list()
+  l$omega$data <- lapply(as.list(omat(x)), get_upper_tri)
+  l$omega$labels <- labels(omat(x))
+  l$omega$names <- names(omat(x))
+  if(length(l$omega$data)) {
+    names(l$omega$data) <- paste0("matrix", seq_along(l$omega$data))
+    names(l$omega$labels) <- paste0("matrix",seq_along(l$omega$labels))
+  }
+  l$sigma <- list()
+  l$sigma$data <- lapply(as.list(smat(x)), get_upper_tri)
+  l$sigma$labels <- labels(smat(x))
+  l$sigma$names <- names(smat(x))
+  if(length(l$sigma$data)) {
+    names(l$sigma$data) <- paste0("matrix", seq_along(l$sigma$data))
+    names(l$sigma$labels) <- paste0("matrix",seq_along(l$sigma$labels))
+  }
+  # Other
+  l$env <- as.list(x@envir)
+  l$plugin <- x@plugin
+  # These items will get directly passed to update()
+  l$update <- list()
+  l$update$start <- x@start
+  l$update$end <- x@end
+  l$update$delta <- x@delta
+  l$update$add <- x@add
+  l$update$atol <- x@atol
+  l$update$rtol <- x@rtol
+  l$update$ss_atol <- x@ss_atol
+  l$update$ss_rtol <- x@ss_rtol
+  l$update$maxsteps <- x@maxsteps
+  l$update$hmax <- x@hmax
+  l$update$hmin <- x@hmin
+  l$update$maxsteps <- x@maxsteps
+  l$update$mxhnil <- x@mxhnil
+  l$update$ixpr <- x@ixpr
+  l$update$mindt <- x@mindt
+  l$update$digits <- x@digits
+  l$update$tscale <- x@tscale
+  l$update$outvars <- unlist(outvars(x), use.names = FALSE)
+  
+  code <- gsub("\\t", "  ", x@code, perl = TRUE)
+  code <- modelparse(code, comment_re = character(0))
+  code <- lapply(code, trimws, which = "right")
+  
+  l$set <- tolist(code$SET)
+  
+  if(nrow(x@annot$data)) {
+    annot <- x@annot$data
+    annot$options <- NULL
+    if(!requireNamespace("knitr", quietly = TRUE)) { 
+      abort("The package \"knitr\" is required.") #nocov
+    } 
+    annot <- knitr::kable(annot, format = "simple")
+    annot <- c("# Model Annotations: ", "", annot)
+    if("PROB" %in% names(code)) {
+      code$PROB <- c(code$PROB, "", annot)  
+    } else {
+      code <- c(list(PROB = annot), code)
+    }
+  }
+  
+  if("PROB" %in% names(code)) {
+    l$prob <- code$PROB
+    code$PROB <- NULL
+  }
+  
+  clob <- c("PARAM", "INPUT", "THETA", "CMT", "INIT", "OMEGA", "SIGMA", 
+            "NMEXT", "NMXML", "VCMT", "SET", "CAPTURE")
+  for(block in clob) {
+    while(block %in% names(code)) {
+      code[[block]] <- NULL
+    }
+  }
```

Comment:
I think this could be simplified to

```R
code[[clob]] <- NULL
```

Reply:
@kyleam can you help me with this one? I'm seeing this:

``` r
clob <- c("OMEGA", "SET", "CMT")

code <- list(a = 5, OMEGA = 3, SET = 4, CMT  = 1, SET = "A", CMT = 3, CMT = 5, 
             b = 4, c = 3)
```

# Not working

``` r
code[clob]
#> $OMEGA
#> [1] 3
#> 
#> $SET
#> [1] 4
#> 
#> $CMT
#> [1] 1
try(code[[clob]])
#> Error in code[[clob]] : recursive indexing failed at level 2
```

# What about this?

``` r
w <- which(names(code) %in% clob)
if(length(w)) {
  code <- code[-w]
}
code
#> $a
#> [1] 5
#> 
#> $b
#> [1] 4
#> 
#> $c
#> [1] 3
```

<sup>Created on 2024-07-23 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
Sorry for the typo :/ I meant `[`.  

But what I didn't appreciate was that there could be multiples.  So either your original code or what you suggest work for me.  Sorry for the noise.

Reply:
Gotcha; thanks. I'll go with the new version since it seems more efficient.